src/nvim/event/loop.c

```diff
@@ -87,6 +87,31 @@ bool loop_poll_events(Loop *loop, int64_t ms)
   return timeout_expired;
 }
 
+/// If `multiqueue` has events then process them and return.
+/// Otherwise if `loop` is non-NULL, run it once and process its fast events.
+///
+/// @param loop
+/// @param multiqueue
+/// @param ms  0: non-blocking poll.
+///            > 0: timeout after `ms`.
+///            < 0: wait forever.
+/// @return  true if no events were processed or timeout expired
+bool loop_process_events(Loop *loop, MultiQueue *multiqueue, int64_t ms)
+{
+  if (multiqueue && !multiqueue_empty(multiqueue)) {
+    multiqueue_process_events(multiqueue);
+    return false;
+  }
+
+  if (loop) {
+    bool timeout_expired = loop_uv_run(loop, ms, true);
+    multiqueue_process_events(loop->fast_events);
+    return timeout_expired;
+  }
+
+  return true;
```

Comment:
This means `vim.wait()` doesn't wait at all in a fast context, which may be confusing.

Reply:
Hmm, doesn't the doc for the `fast_only` parameter imply that this is already the case? 

Do we need a `fast_loop` where non-editor, Lua-only processing is done?

```
      • {fast_only}  (boolean|nil) If true, only |api-fast| events will be
                     processed. If called from while in an |api-fast| event,
                     will automatically be set to `true`.
```

Reply:
One of the main use cases for `vim.wait()` is to for applications like `vim.system()` where we want to uv to poll and check for IO, however this only happens on the main loop.

> Or does NULL already work like that?

NULL just ensures we never run the main loop, but it will still process any outstanding fast/main events. However I think zeer is right and maybe this change is wrong.

I don't see how we can run the uv loop when we are in a fast callback, so maybe we just disallow it since it would become a massive foot gun otherwise.

> Do we need a fast_loop where non-editor, Lua-only processing is done? 

This probably surpasses my knowledge of how the event loop works, but are you suggesting we create another loop which uv binds to in fast-callbacks? Could that even work?

Reply:
IMO introducing a secondary event loop would be the wrong way to go. 

if we want to allow "sync style" code inside callback (i e a `vim.system` which waits and returns, instead of taking yet another callback), we need to use something based on co-routines, so the callback can be suspended and the top-level event loop resumed.

Reply:
> I don't see how we can run the uv loop when we are in a fast callback,

Another (half-baked, probably wrong) thought was for the fast callback to "re-schedule itselfon the main loop". But suspend-and-resume as @bfredl mentioned, sounds much better.

Reply:
> It may just to easier to disallow `vim.wait` in fast events.

Throwing an error is fine for now, because it's a 2-way door: we can always relax it in the future.